<launch>
  <arg name="map_size_x_"/>
  <arg name="map_size_y_"/>
  <arg name="map_size_z_"/>
  <arg name="frame_width"/>
  <arg name="frame_height"/>
  <arg name="odometry_topic"/>
  <arg name="sensor_pose_topic"/>
  <arg name="depth_topic"/>
  <arg name="cloud_topic"/>
  <arg name="cx"/>
  <arg name="cy"/>
  <arg name="fx"/>
  <arg name="fy"/>
  <arg name="max_vel"/>
  <arg name="max_acc"/>
  <arg name="point_num"/>
  <arg name="point0_x"/>
  <arg name="point0_y"/>
  <arg name="point0_z"/>
  <arg name="point1_x"/>
  <arg name="point1_y"/>
  <arg name="point1_z"/>
  <arg name="point2_x"/>
  <arg name="point2_y"/>
  <arg name="point2_z"/>
  <arg name="flight_type"/>
  <arg name="clearance"    default="0.9"/>
  <arg name="z_height_max" default="2.0"/>
  <arg name="z_height"     default="$(eval arg('z_height_max') + arg('clearance'))"/>
  <arg name="log_output"   default="screen"/>
  <arg name="enable_viz"   default="true"/>
  <arg name="depth_dist"   default="3.0"/> <!-- Максимальная дальность измерения камерой глубины -->
  
  <!-- main node -->
  <node pkg="plan_manage" name="fast_planner_node" type="fast_planner_node" output="$(arg log_output)">
    <remap from="/odom_world"       to="$(arg odometry_topic)"/>
    <remap from="/map_ros/pose"     to = "$(arg sensor_pose_topic)"/> 
    <remap from="/map_ros/depth"    to = "$(arg depth_topic)"/>
    <remap from="/map_ros/cloud"    to="$(arg cloud_topic)"/>
    <remap from="/planning/replan"  to="planning/replan"/>
    <remap from="/planning/new"     to="planning/new"/>
    <remap from="/planning/bspline" to="planning/bspline"/>
    <remap from="/planning/wait"    to="planning/wait"/>
    <remap from="/waypoint_generator/waypoints" to="~waypoints"/>
    
    <!-- replanning method -->
    <param name="planner_node/planner" value="2" type="int"/>

    <!-- planning fsm -->
    <param name="fsm/flight_type" value="$(arg flight_type)" type="int"/>
    <param name="fsm/thresh_replan" value="0.5" type="double"/>
    <param name="fsm/thresh_no_replan" value="1.0" type="double"/>
    <param name="fsm/act_map" value="false" type="bool"/>

    <param name="fsm/waypoint_num" value="$(arg point_num)" type="int"/>
    <param name="fsm/waypoint0_x" value="$(arg point0_x)" type="double"/>
    <param name="fsm/waypoint0_y" value="$(arg point0_y)" type="double"/>
    <param name="fsm/waypoint0_z" value="$(arg point0_z)" type="double"/>
    <param name="fsm/waypoint1_x" value="$(arg point1_x)" type="double"/>
    <param name="fsm/waypoint1_y" value="$(arg point1_y)" type="double"/>
    <param name="fsm/waypoint1_z" value="$(arg point1_z)" type="double"/>
    <param name="fsm/waypoint2_x" value="$(arg point2_x)" type="double"/>
    <param name="fsm/waypoint2_y" value="$(arg point2_y)" type="double"/>
    <param name="fsm/waypoint2_z" value="$(arg point2_z)" type="double"/>

    <param name="sdf_map/resolution"   value="0.1" /> 
    <param name="sdf_map/frame_width"  value="$(arg frame_width)" /> 
    <param name="sdf_map/frame_height" value="$(arg frame_height)" /> 
    <param name="sdf_map/map_size_x"   value="$(arg map_size_x_)" /> 
    <param name="sdf_map/map_size_y"   value="$(arg map_size_y_)" /> 
    <param name="sdf_map/map_size_z"   value="$(arg map_size_z_)" /> 
    <param name="sdf_map/obstacles_inflation" value="0.15" /> 
    <param name="sdf_map/local_bound_inflate" value="0.5"/> <!-- Облась обработки 3д карты (update range) -->
    <param name="sdf_map/local_map_margin"    value="25"/>  <!-- Размер esdf карты -->
    <param name="sdf_map/ground_height"       value="0.3"/>
    <param name="sdf_map/default_dist"        value="0"/> <!-- Расстояние до недостигнутого вокселя на sdf -->

    <param name="sdf_map/p_hit"  value="0.65"/>
    <param name="sdf_map/p_miss" value="0.43"/>
    <param name="sdf_map/p_min"  value="0.12"/>
    <param name="sdf_map/p_max"  value="0.90"/>
    <param name="sdf_map/p_occ"  value="0.80"/>
    <param name="sdf_map/min_ray_length" value="0.3"/>
    <param name="sdf_map/max_ray_length" value="$(arg depth_dist)"/>
    <param name="sdf_map/virtual_ceil_height"   value="$(arg z_height)"/>
    <param name="sdf_map/optimistic" value="true" type="bool"/>
    <param name="sdf_map/signed_dist" value="true" type="bool"/>

    <param name="map_ros/cx" value="$(arg cx)"/>
    <param name="map_ros/cy" value="$(arg cy)"/>
    <param name="map_ros/fx" value="$(arg fx)"/>
    <param name="map_ros/fy" value="$(arg fy)"/>
    <param name="map_ros/depth_filter_maxdist"   value="$(eval 0.5 + arg('depth_dist'))"/>
    <param name="map_ros/depth_filter_mindist"   value="0.2"/>
    <param name="map_ros/depth_filter_margin"    value="2"/>
    <param name="map_ros/k_depth_scaling_factor" value="1000.0"/>
    <param name="map_ros/skip_pixel" value="1"/>
    <param name="map_ros/esdf_slice_height" value="1.0"/>
    <param name="map_ros/visualization_truncate_height"   value="$(arg z_height)"/>
    <param name="map_ros/visualization_truncate_low"   value="0.0"/>
    <param name="map_ros/frame_id"       value="world"/>
    <param name="map_ros/show_occ_time"  value="false"/>
    <param name="map_ros/show_esdf_time" value="false"/>
    <param name="map_ros/enable_vis"     value="$(arg enable_viz)"/>
    <param name="map_ros/show_all_map"   value="$(arg enable_viz)"/>

  <!-- planner manager -->
    <param name="manager/min_time" value="true" type="bool"/>
    <param name="manager/max_vel"  value="$(arg max_vel)" type="double"/>
    <param name="manager/max_acc"  value="$(arg max_acc)" type="double"/>
    <param name="manager/max_jerk" value="4" type="double"/>
    <param name="manager/accept_vel" value="$(eval 0.0+arg('max_vel'))" type="double"/>
    <param name="manager/accept_acc" value="$(eval 0.0+arg('max_acc'))" type="double"/>
    <param name="manager/dynamic_environment"  value="0" type="int"/>
    <param name="manager/clearance_threshold"  value="$(arg clearance)" type="double"/>
    <param name="manager/local_segment_length" value="2.5" type="double"/>
    <param name="manager/control_points_distance" value="0.5" type="double"/>

    <param name="manager/use_geometric_path" value="false" type="bool"/>
    <param name="manager/use_kinodynamic_path" value="false" type="bool"/>
    <param name="manager/use_topo_path" value="true" type="bool"/>
    <param name="manager/use_optimization" value="true" type="bool"/>
    <param name="manager/use_active_perception" value="false" type="bool"/>

  <!-- topology path finding -->
    <param name="topo_prm/parallel_shortcut" value="false" type="bool"/>
    <param name="topo_prm/sample_inflate_x" value="1.5" type="double"/>
    <param name="topo_prm/sample_inflate_y" value="3.5" type="double"/>
    <param name="topo_prm/sample_inflate_z" value="0.5" type="double"/>
    <param name="topo_prm/max_sample_time" value="0.005" type="double"/>
    <param name="topo_prm/max_sample_num"  value="2000" type="int"/>
    <param name="topo_prm/clearance"       value="$(arg clearance)" type="double"/>
    <param name="topo_prm/max_raw_path"    value="300" type="int"/>
    <param name="topo_prm/max_raw_path2"   value="25" type="int"/>
    <param name="topo_prm/short_cut_num"   value="1" type="int"/>
    <param name="topo_prm/reserve_num"     value="6" type="int"/>
    <param name="topo_prm/ratio_to_short"  value="5.5" type="double"/>

  <!-- trajectory optimization -->
    <param name="optimization/ld_smooth" value="20.0" type="double"/>
    <param name="optimization/ld_dist" value="10.0" type="double"/>
    <param name="optimization/ld_feasi" value="1.0" type="double"/>
    <param name="optimization/ld_start" value="10.0" type="double"/>
    <param name="optimization/ld_end" value="1.0" type="double"/>
    <param name="optimization/ld_guide" value="2" type="double"/>
    <param name="optimization/ld_waypt" value="5.0" type="double"/>
    <param name="optimization/ld_view" value="0.0" type="double"/>
    <param name="optimization/ld_time" value="0.2" type="double"/>
    
    <param name="optimization/dist0" value="0.65" type="double"/>
    <param name="optimization/dist1" value="0.0" type="double"/>
    <param name="optimization/max_vel" value="$(arg max_vel)" type="double"/>
    <param name="optimization/max_acc" value="$(arg max_acc)" type="double"/>

    <param name="optimization/algorithm1" value="15" type="int"/>
    <param name="optimization/algorithm2" value="11" type="int"/>
    <param name="optimization/max_iteration_num1" value="2" type="int"/>
    <param name="optimization/max_iteration_num2" value="1000" type="int"/>
    <param name="optimization/max_iteration_num3" value="2000" type="int"/>
    <param name="optimization/max_iteration_time1" value="0.0001" type="double"/>
    <param name="optimization/max_iteration_time2" value="0.005" type="double"/>
    <param name="optimization/max_iteration_time3" value="0.010" type="double"/>
    <param name="bspline/limit_vel" value="$(arg max_vel)" type="double"/>
    <param name="bspline/limit_acc" value="$(arg max_acc)" type="double"/>
    <param name="bspline/limit_ratio" value="1.1" type="double"/>

    <param name="heading_planner/yaw_diff" value="$(eval 5 * 3.1415926 / 180.0)" type="double"/>
    <param name="heading_planner/half_vert_num" value="3" type="int"/>
    <param name="heading_planner/lambda1" value="2.0" type="double"/>
    <param name="heading_planner/lambda2" value="1.0" type="double"/>
    <param name="heading_planner/max_yaw_rate" value="$(eval 60 * 3.1415926 / 180.0)" type="double"/>
    <param name="heading_planner/w" value="10.0" type="double"/>
    <param name="heading_planner/weight_type" value="0" type="double"/>
    <!-- <param name="fsm/act_map" value="true" type="bool"/> -->
    <param name="fsm/act_map" value="false" type="bool"/>

    <param name="optimization/wnl" value="1.0" type="double"/>
    <param name="visibility/visib_min" value="0.25" type="double"/>
    <param name="visibility/max_acc" value="$(arg max_acc)" type="double"/>
    <param name="visibility/r0" value="0.0" type="double"/>
    <param name="visibility/forward" value="0.5" type="double"/>
    <param name="visibility/max_safe_dist" value="2.0" type="double"/>
    <param name="visibility/safe_margin" value="0.5" type="double"/>
  </node>

</launch>